### PR TITLE
pure virtual functinos overriden

### PR DESCRIPTION
**Pull request by D300**

### CODE REVIEW

File: AttackItem.h [hunk 1]
```diff
@@ -1 +1,16 @@
-#include "Item.h"
\ No newline at end of file
+#pragma once
+
+#include "AttackItem.h"
+#include "Item.h"
+#include "Utility.h"
+
```
Comment (matkatmusic) [inline AttackItem.h:6]: You're not using Utility.h here, but you are using it in the cpp file.
- [ ] include only what you use.

File: Character.cpp [hunk 2]
```diff
@@ -86,8 +92,16 @@ int Character::takeDamage(int damage)
...
 }
 
+void boost(int& current, int& initial)
+{
+    if ( current < initial)
+            current = initial;
+        
+        current *= 1.1f;
+        initial = current;
+}
```
Comment (matkatmusic) [inline Character.cpp:102]: Formatting
- [ ] revise

File: Character.h [hunk 3]
```diff
@@ -7,14 +7,15 @@
 
 #include "Item.h"
 
+//ctor of dwarf: name, hit points, armor.
 struct Character
 {
     Character(int hp, int armor_, int attackDamage_ );
-    virtual ~Character() { }
+    virtual ~Character(); // must be overriden?
```
Comment (matkatmusic) [inline Character.h:14]: no, it's the base class.  subclasses override it.

File: DefensiveItem.cpp [hunk 4]
```diff
@@ -1,12 +1,10 @@
...
 void DefensiveItem::use(Character* character)
 {
-    assert(false); 
-    //make your defensive item use the appropriate Utility helper functions
-}
\ No newline at end of file
+    // why derefencing this?
+    useDefensiveItem(character, *this);
+}
+
```
Comment (matkatmusic) [inline DefensiveItem.cpp]: look at the signature for `useDefensiveItem`, 2nd param.

File: Dwarf.cpp [hunk 5]
```diff
@@ -1,7 +1,25 @@
 #include "Dwarf.h"
+#include "Utility.h"
 
-//Dwarf::Dwarf
+// base class ctor:
+// does attackDamage equal armor?
+// ! Character(int hp, int armor_, int attackDamage_ );
```
Comment (matkatmusic) [inline Dwarf.cpp:6]: ??  This is a question you shouldn't be asking...

File: Utility.h [hunk 6]
```diff
@@ -4,6 +4,8 @@
 #include <memory>
 #include "Item.h"
 
+// what is this - funcs without any class?
+// would it make sense to make them static then?
```
Comment (matkatmusic) [inline Utility.h:8]: No, they are free functions.  It's simpler to make them free functions, vs putting them in a `Utilities` namespace, or writing a `Utilities` class and making them all `public static`. Both options require adding a `Utilities::` qualifier before calling them.  Nothing is gained by doing this.

File: main.cpp [hunk 7]
```diff
@@ -56,14 +56,36 @@ send me a DM to check your pull request
...
+
+
+main.cpp:(.text+0x20b): undefined reference to `Character::defend()'
+
+/tmp/main-4061ea.o: In function `Character::~Character()':
+main.cpp:(.text._ZN9CharacterD2Ev[_ZN9CharacterD2Ev]+0xa): undefined reference to `vtable for Character'
+
+clang-7: error: linker command failed with exit code 1 (use -v to see invocation)
+
+*/
```
Comment (matkatmusic) [inline main.cpp:74]: These errors are gone, so you should delete these notes to yourself.

File: AttackItem.h (hunk 1 above)
Comment (D300) [reply AttackItem.h:6]: in JUCE I started to include always only in the header files, so there are less places to look at.. makes it more sorted to me. what do you think about this?

File: Character.h (hunk 3 above)
Comment (D300) [reply Character.h:14]: but none of the childclasses (== subclasses ?) is deleted anytime, so the virtual destructor doesnt come into play in that code right?

File: DefensiveItem.cpp (hunk 4 above)
Comment (D300) [reply DefensiveItem.cpp]: jep I understand this forgot to delete the question

File: Dwarf.cpp (hunk 5 above)
Comment (D300) [reply Dwarf.cpp:6]: jep I understand this forgot to delete the question

**Pull request merged**